File: Algorithms/Java/Generate Parentheses.java
public class Solution {
    public static ArrayList<String> generateParenthesis(int n) {
		ArrayList<String> ans = new ArrayList<String>();
		getPairs(ans, "", n, n);
		return ans;
	}

	public static void getPairs(ArrayList<String> ans, String s, int left,
			int right) {
		if (left > right || left < 0 || right < 0)
			return;
		if (left == 0 && right == 0) {
			ans.add(s);
			return;
		}
		getPairs(ans, s + "(", left - 1, right);
		getPairs(ans, s + ")", left, right - 1);
	}
}

File: Algorithms/Python/Generate-Parentheses.py
class Solution:
    # @param an integer
    # @return a list of string

    def generateParenthesis(self, n):

        def dfs(cur, left, right):
            if left == 0 and right == 0:
                ans.append(cur)
                return
            if left > right or left < 0 or right < 0:
                return
            dfs(cur + '(', left - 1, right)
            dfs(cur + ')', left, right - 1)

        ans = []
        dfs('', n, n)
        return ans
